src/program/config/listen/listen.lua

```diff
@@ -0,0 +1,318 @@
+      -- ourselves to printable ASCII characters.
+      local part = take_while(input, literal_string_chars_pat)
+      if part == '' then
+         consume(input, "\\")
+         for k,v in pairs(escaped_string_chars) do
+            if check(input, k) then part = v; break end
+         end
+         if part == '' and check(input, "u") then
+            -- 4-hex-digit unicode escape.  We only support ASCII
+            -- tho.
+            local hex = '0x'
+            for i=1,4 do hex = hex..consume_pat(input, "%x") end
+            local code = assert(tonumber(hex))
+            if code >= 128 then error('non-ASCII character: \\u00'..hex) end
+            part = string.char(code)
+         end
+      end
+      table.insert(parts, part)
+   end
+   return table.concat(parts)
+end
+
+local function read_json_object(input)
+   consume(input, "{")
+   skip_whitespace(input)
+   local ret = {}
+   if not check(input, "}") then
+      repeat
+         skip_whitespace(input)
+         local k = read_json_string(input)
+         if ret[k] then error('duplicate key: '..k) end
+         skip_whitespace(input)
+         consume(input, ":")
+         skip_whitespace(input)
+         local v = read_json_string(input)
+         ret[k] = v
+         skip_whitespace(input)
+      until not check(input, ",")
+      skip_whitespace(input)
+      consume(input, "}")
+   end
+   return ret
+end
+
+local function buffered_output()
+   local ret = { buf = {} }
+   function ret:write(str) table.insert(self.buf, str) end
+   function ret:flush_to_fd(fd)
+      local str = table.concat(self.buf)
+      local bytes = ffi.new('uint8_t[?]', #str, str)
+      local written = 0
+      while written < #str do
+         local wrote = assert(S.write(fd, bytes + written, #str - written))
+         written = written + wrote
+      end
+   end
+   return ret
+end
+
+local function write_json_string(output, str)
+   output:write('"')
+   local pos = 1
+   while pos <= #str do
+      local head = str:match('^('..literal_string_chars_pat..'+)', pos)
+      if head then
+         output:write(head)
+         pos = pos + #head
+      else
+         head = str:sub(pos, pos)
+         local escaped
+         for k,v in pairs(escaped_string_chars) do
+            if v == head then escaped = k; break end
+         end
+         if not escaped then
+            escaped = string.format("u00%.2x", head:byte(1))
+         end
+         output:write('\\'..escaped)
+         pos = pos + 1
+      end
+   end
+   output:write('"')
+end
+
+local function write_json_object(output, obj)
+   output:write('{')
+   local comma = false
+   for k,v in pairs(obj) do
+      if comma then output:write(',') else comma = true end
+      write_json_string(output, k)
+      output:write(':')
+      write_json_string(output, v)
+   end
+   output:write('}')
+end
+
+local function validate_value(schema_name, revision_date, path, value_str)
+   local parser = common.data_parser(schema_name, path)
```

Comment:
This is a general comment that applies to all of our yang library :)  I think this is one of those things that will fix itself in time, as we add revisions to yang schemas.  Is it a blocker for you?  LMK.